File: container/skills/openrouter-image-analysis/scripts/analyze_image.mjs
#!/usr/bin/env node

/**
 * Analyze an image using OpenRouter's vision models.
 *
 * Zero external dependencies — uses only Node.js built-ins.
 * Accepts local file paths (png, jpg, jpeg, webp, gif, bmp) or HTTP/HTTPS URLs.
 */

import { readFileSync, existsSync } from 'fs';
import { extname, resolve, dirname } from 'path';
import { execFileSync } from 'child_process';
import { fileURLToPath } from 'url';

const OPENROUTER_API_URL = 'https://openrouter.ai/api/v1/chat/completions';

const DEFAULT_PROMPT =
  'Describe the image and extract any important visible text, errors, UI elements, diagrams, or notable details.';

const SUPPORTED_EXTENSIONS = new Set([
  '.png',
  '.jpg',
  '.jpeg',
  '.webp',
  '.gif',
  '.bmp',
]);

const MIME_MAP = {
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.webp': 'image/webp',
  '.gif': 'image/gif',
  '.bmp': 'image/bmp',
};

function die(msg) {
  process.stderr.write(`Error: ${msg}\n`);
  process.exit(1);
}

function encodeLocalImage(filePath) {
  const abs = resolve(filePath.replace(/^~/, process.env.HOME || '~'));
  if (!existsSync(abs)) die(`file not found: ${abs}`);

  const ext = extname(abs).toLowerCase();
  if (!SUPPORTED_EXTENSIONS.has(ext)) {
    die(
      `unsupported image format '${ext}'. Supported: ${[...SUPPORTED_EXTENSIONS].sort().join(', ')}`,
    );
  }

  const mime = MIME_MAP[ext] || `image/${ext.slice(1)}`;
  const data = readFileSync(abs);
  const b64 = Buffer.from(data).toString('base64');
  return `data:${mime};base64,${b64}`;
}

function resolveModel() {
  const pinned = process.env.OPENROUTER_VISION_MODEL;
  if (pinned) return pinned;

  const __dirname = dirname(fileURLToPath(import.meta.url));
  const helper = resolve(__dirname, 'find_free_models.mjs');

  if (!existsSync(helper)) die(`helper script not found: ${helper}`);

  let stdout;
  try {
    stdout = execFileSync('node', [helper], {
      encoding: 'utf-8',
      timeout: 30_000,
    });
  } catch (e) {
    if (e.code === 'ENOENT')
      die(
        "'node' not found in PATH. Set OPENROUTER_VISION_MODEL to skip discovery.",
      );
    if (e.killed) die('model discovery timed out after 30s');
    die(`model discovery failed: ${(e.stderr || '').trim()}`);
  }

  const models = stdout
    .split('\n')
    .map((l) => l.trim())
    .filter(Boolean);

  if (models.length === 0) {
    die(
      'no free vision-capable models found on OpenRouter. Set OPENROUTER_VISION_MODEL to specify one manually.',
    );
  }

  return models[0];
}

async function analyzeImage(imageRef, prompt, model) {
  const apiKey = process.env.OPENROUTER_API_KEY;
  if (!apiKey) die('OPENROUTER_API_KEY environment variable not set');

  const imageUrl =
    imageRef.startsWith('http://') || imageRef.startsWith('https://')
      ? imageRef
      : encodeLocalImage(imageRef);

  const payload = {
    model,
    messages: [
      {
        role: 'user',
        content: [
          { type: 'text', text: prompt },
          { type: 'image_url', image_url: { url: imageUrl } },
        ],
      },
    ],
    max_tokens: 4096,
  };

  let resp;
  try {
    resp = await fetch(OPENROUTER_API_URL, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(120_000),
    });
  } catch (e) {
    die(`request failed: ${e.message}`);
  }

  if (!resp.ok) {
    const body = await resp.text().catch(() => '');
    die(`OpenRouter API returned ${resp.status}: ${body.slice(0, 500)}`);
  }

  let data;
  try {
    data = await resp.json();
  } catch {
    die('failed to parse API response as JSON');
  }

  try {
    return data.choices[0].message.content;
  } catch {
    die(`unexpected API response: ${JSON.stringify(data).slice(0, 500)}`);
  }
}

// --- main ---

const args = process.argv.slice(2);
if (args.length < 1) {
  process.stderr.write(
    'Usage: analyze_image.mjs <image_path_or_url> [prompt]\n' +
      '  image_path_or_url  local file path or http(s) URL\n' +
      '  prompt             optional analysis instruction\n',
  );
  process.exit(1);
}

const imageRef = args[0];
const prompt = args.slice(1).join(' ') || DEFAULT_PROMPT;
const model = resolveModel();

process.stderr.write(`[using model: ${model}]\n`);

const result = await analyzeImage(imageRef, prompt, model);
process.stdout.write(result + '\n');

File: container/skills/openrouter-image-analysis/scripts/analyze_image.py
#!/usr/bin/env python3
"""Analyze an image using OpenRouter's vision models.

Zero external dependencies — uses only Python stdlib.
Accepts local file paths (png, jpg, jpeg, webp, gif) or HTTP/HTTPS URLs.
"""

import sys
import os
import base64
import mimetypes
import json
import subprocess
import urllib.request
import urllib.error

OPENROUTER_API_URL = "https://openrouter.ai/api/v1/chat/completions"

DEFAULT_PROMPT = (
    "Describe the image and extract any important visible text, "
    "errors, UI elements, diagrams, or notable details."
)

SUPPORTED_EXTENSIONS = {".png", ".jpg", ".jpeg", ".webp", ".gif", ".bmp"}


def die(msg):
    print(f"Error: {msg}", file=sys.stderr)
    sys.exit(1)


def encode_local_image(path):
    """Encode a local image file as a base64 data URL."""
    path = os.path.expanduser(path)
    if not os.path.isfile(path):
        die(f"file not found: {path}")

    ext = os.path.splitext(path)[1].lower()
    if ext not in SUPPORTED_EXTENSIONS:
        die(f"unsupported image format '{ext}'. Supported: {', '.join(sorted(SUPPORTED_EXTENSIONS))}")

    mime_type = mimetypes.guess_type(path)[0]
    if not mime_type:
        mime_type = f"image/{ext.lstrip('.')}"

    with open(path, "rb") as f:
        encoded = base64.b64encode(f.read()).decode("utf-8")

    return f"data:{mime_type};base64,{encoded}"


def resolve_model():
    """Resolve which vision model to use.

    Priority: OPENROUTER_VISION_MODEL env var > auto-discover via helper script.
    """
    model = os.getenv("OPENROUTER_VISION_MODEL")
    if model:
        return model

    script_dir = os.path.dirname(os.path.abspath(__file__))
    helper = os.path.join(script_dir, "find_free_models.mjs")

    if not os.path.isfile(helper):
        die(f"helper script not found: {helper}")

    try:
        result = subprocess.run(
            ["node", helper],
            capture_output=True,
            text=True,
            timeout=30,
        )
    except FileNotFoundError:
        die("'node' not found in PATH. Install Node.js or set OPENROUTER_VISION_MODEL.")
    except subprocess.TimeoutExpired:
        die("model discovery timed out after 30s")

    if result.returncode != 0:
        die(f"model discovery failed: {result.stderr.strip()}")

    models = [line.strip() for line in result.stdout.strip().splitlines() if line.strip()]
    if not models:
        die(
            "no free vision-capable models found on OpenRouter. "
            "Set OPENROUTER_VISION_MODEL to specify one manually."
        )

    return models[0]


def analyze_image(image_ref, prompt, model):
    """Send image to OpenRouter for analysis, return plain text."""
    api_key = os.getenv("OPENROUTER_API_KEY")
    if not api_key:
        die("OPENROUTER_API_KEY environment variable not set")

    # Build image content block
    if image_ref.startswith("http://") or image_ref.startswith("https://"):
        image_url = image_ref
    else:
        image_url = encode_local_image(image_ref)

    payload = {
        "model": model,
        "messages": [
            {
                "role": "user",
                "content": [
                    {"type": "text", "text": prompt},
                    {"type": "image_url", "image_url": {"url": image_url}},
                ],
            }
        ],
        "max_tokens": 4096,
    }

    body = json.dumps(payload).encode("utf-8")

    req = urllib.request.Request(
        OPENROUTER_API_URL,
        data=body,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            response_data = json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        error_body = e.read().decode("utf-8", errors="replace")[:500]
        die(f"OpenRouter API returned {e.code}: {error_body}")
    except urllib.error.URLError as e:
        die(f"network error: {e.reason}")
    except Exception as e:
        die(f"request failed: {e}")

    try:
        return response_data["choices"][0]["message"]["content"]
    except (KeyError, IndexError, TypeError):
        die(f"unexpected API response: {json.dumps(response_data)[:500]}")


def main():
    if len(sys.argv) < 2:
        print(
            "Usage: analyze_image.py <image_path_or_url> [prompt]\n"
            "  image_path_or_url  local file path or http(s) URL\n"
            "  prompt             optional analysis instruction",
            file=sys.stderr,
        )
        sys.exit(1)

    image_ref = sys.argv[1]
    prompt = " ".join(sys.argv[2:]) if len(sys.argv) > 2 else DEFAULT_PROMPT

    model = resolve_model()

    # Print model to stderr so stdout stays clean for programmatic consumers
    print(f"[using model: {model}]", file=sys.stderr)

    result = analyze_image(image_ref, prompt, model)
    print(result)


if __name__ == "__main__":
    main()
